Add an ingest event audit logger

djjuhasz opened the pull request:
Refs #1282.

- Add an ingest audit logger that subscribes to the ingest event service and writes "SIP deposited" events to an audit log file
- Configure the development environment to do audit logging

djjuhasz committed:
Configure audit logging in dev environment

- Switch from the Docker distroless/base-debian12:latest to
  debian:13-slim as base image for enduro containers to allow shell
  access and running `mkdir` to create a log directory
- Create a `/home/enduro/logs/` directory in the enduro container
- Configure enduro to log to `/home/enduro/logs/enduro_ingest_audit.log`

djjuhasz commented:
Here's what the audit log output looks like right now:
```
{"time":"2025-08-18T20:37:20.622422365Z","level":"INFO","msg":"SIP deposited","type":"SIP.deposit","objectID":"bc70d029-8fa6-4455-9523-739f93fbe8a1","userID":"44675334-8d13-474f-8954-4db06a393d65"}
{"time":"2025-08-18T21:51:28.115995153Z","level":"INFO","msg":"SIP deposited","type":"SIP.deposit","objectID":"ed225ceb-e7a5-47ce-a2e9-871b32ac2879","userID":"44675334-8d13-474f-8954-4db06a393d65"}
{"time":"2025-08-18T21:51:28.150393933Z","level":"INFO","msg":"SIP deposited","type":"SIP.deposit","objectID":"1778a8b2-beb0-4cdd-8b7c-3aee703776b9","userID":"44675334-8d13-474f-8954-4db06a393d65"}
```

djjuhasz commented:
@sevein if you have time to review this PR sometime this week, I'd appreciate it.  If not then @jraddaoui can review it next week when he's back from vacation.

sevein reviewed (commented):
In regards to the changes to the Dockerfile, it looks like a significant amount of effort has gone into creating a secure prod-ready image based on distroless, and this change feels like undoing that work for the sake of developer ergonomics.

Perhaps we could explore other alternatives? We could either maintain a separate, development-focused Dockerfile build stage that produces an image with a shell (feels undesirable), or perhaps we could address the `mkdir` and shell access concerns in a way that doesn't require changing the base image. For instance, the log directory can be created in a previous multi-stage build and then copied into the final distroless image. For shell access, we can adopt the use of ephemeral containers for debugging purposes - I remember we considered this option before and it had some drawbacks but seemed good enough to preserve the benefits of the distroless image while providing a good dev experience?

sevein reviewed (commented):
Here is some feedback on the new audit log implementation:

* The `ingest.Service` interface is being polluted by the `WithAuditLogger` method because is not a core responsibility of the ingest service but a setup concern. I would encourage instead to continue using the pattern of passing dependencies to the `ingest.NewService` constructor. You could adopt the "functional options" pattern (e.g. pass WithAuditLogger to the constructor) but it doesn't look like a good use case to me.
* The is no feedback if the audit logger fails to subscribe to the event bus on startup. We could alter the signature of `ingest.NewService` constructor to return such error, allowing the app to fail if the audit system can't be initialized.
* I'm wondering if the audit log's subscription logic should be moved instead to the dedicated `auditlog` package, i.e. a package where we address all aspects of audit logging including subscription details? Not entirely sure.
* It feels like the audit logging system should have a Shutdown method because you'd want to have a mechanism to wait for the logging goroutine to complete successfully during shutdown. Perhaps something that can be implemented with a `sync.WaitGroup` ensuring that the main application does not exit until all the pending audit events have been processed.

Overall I think that the auditlog pkg should contain the audit logging service, not just its config. The subscription and graceful shutdown mechanisms seem audit-related operational concerns as opposed to core ingest business logic.

djjuhasz commented:
@sevein we previously discussed using ephemeral containers and switching from a distroless container in this PR: https://github.com/artefactual-sdps/enduro/pull/1005.

I think at this point we should consider using a distroless distribution for our Docker build a failed experiment and discontinue its use because its disadvantages outweigh its advantages.  We switched to the current distroless Linux distrubtion from an Alpine Linux distribution on a suggestion from you after we introduced the https://github.com/artefactual-labs/bagit-gython dependency, which won't run in musl based Linux distribution.  I chose to use distroless without any knowledge of the drawbacks for development, and without any thorough analysis of the tradeoffs.  

Now that I've tried developing with a distroless container I've found it a significant obstacle to quick and iterative development due to its lack of standard binaries we depend on like libxml and ssh/sftp, and its lack of a shell for inspecting the running container and allowing file access.  Furthermore we don't use this Docker container for any production deployments of Enduro so the security benefits of using a distroless container are negligible in my opinion.  If we do want to create a secure, production-ready, Enduro image I think we should have a dedicated Dockerfile or build system for doing so, and take a more holistic approach to security for the entire Enduro system (e.g. I've used shell access to an Archivematica installation to get access to a PIP created by Enduro).

In the current case the distroless container caused several issues that made it harder to develop this audit logging feature:
1. I could not easily create a new directory for the logs and give the Enduro user write access to the directory to create a log file, due to the lack of standard Linux commands like `mkdir` and `chmod` in the image.
2. I could not easily access the log file written to the container to check that the logger was functioning properly

In the first case, it looks like we could create an empty `/logs` directory and copy it to the distroless image, which is only slightly more cumbersome than just creating the directory directly but it wasn't immediately obvious to me as a solution.  In the second case, we have had no success in our several attempts to access files inside a distroless container using ephemeral containers or other methods.  Ephemeral containers are not easy to understand or use, and thus present an significant barrier to development even if we can get them to work for this purpose.

For all these reasons, I think we should switch back to a standard Linux distribution with shell access for our Enduro Docker file.

djjuhasz commented:
@sevein I've tried several times to create a pure "auditlog" package, but the problem with creating an event listener is the events emitted from an event subscription are typed, and the events we publish have no common interface that would allow me to go from at typed event like `*goaingest.SIPCreatedEvent` to a `[]string` or struct that I could pass to `slog.Info(msg, args...)`.  I'm trying to wrap my head around how to create an adapter to get the typed subscription like `Subscription[T *goaingest.IngestEvent]` to emit a general `auditlog.Event` event, but I haven't been able to come up with a solution yet. :(

sevein commented:
> For all these reasons, I think we should switch back to a standard Linux distribution with shell access for our Enduro Docker file.

Makes sense, thank you!

djjuhasz committed:
Add an ingest event audit logger

Refs #1282.

Add an ingest event service listener that logs SIP creation events to a
log file.

- Use the go `log/slog` standard package to provide a structured logger
  than can be easily configured to write JSON events to a log file
- Add an audit logger to the enduro worker
- Add a handler to convert goaingest SIP deposit events to an
  `auditlog.Event` for logging (more event types to be added later)
- Implement log rotation using the gopkg.in/natefinch/lumberjack.v2
  package

djjuhasz commented:
@sevein and @jraddaoui I've completely restructured the audit logger based on @sevein's feedback to decouple it from the ingest service altogether. I've used a event handler function to allow general conversion of an event service event to an `auditlog.Event` that can then be logged by slog.  

@sevein I don't know if you want to continue code review on this, or if you want to pass it over to @jraddaoui as he's back on Monday.  If you do continue, then please note that I overwrote the previous commit history because this was such a major change — sorry if that makes it harder to review again.

djjuhasz commented:
Oh, I've also incorporated log rotation into this PR because it's a relatively small amount of code, and to avoid having to do a big gnarly rebase of https://github.com/artefactual-sdps/enduro/pull/1324.

jraddaoui commented on cmd/enduro/main.go:
```diff
@@ -164,6 +165,20 @@ func main() {
 	}
 
+	// Set up the ingest audit logger, if one is configured.
+	if cfg.AuditLog.Filepath != "" {
```
We are not really consistent with this, but in other configurations we have an "Enabled" boolean. Not a big deal but I kind of like that pattern.

djjuhasz replied:
I've been avoiding an "enabled" boolean on the principle of avoiding an unnecessary variables — the audit log won't work without a valid file path so toggling the feature based on the presence of a file path seems alternative.  I'm get that an explicit "enabled" boolean needs less explanation though, so I'm fine to add it if you want to go that way.

jraddaoui replied:
Not a big deal, it's okay like it is.

jraddaoui commented on cmd/enduro/main.go:
```diff
@@ -164,6 +165,20 @@ func main() {
+	// Set up the ingest audit logger, if one is configured.
+	if cfg.AuditLog.Filepath != "" {
+		al := auditlog.NewFromConfig(cfg.AuditLog, ingest.HandleAuditEvent)
+		err := al.Listen(ctx, ingestEventSvc)
```
I don't think we should tie the audit logs with the current event services used for the WebSocket updates. Those are meant to send updates to the dashboard so they may not include some details, or we may be missing some of the events needed. For example, the SIP creation event is published once from the processing workflow without user details, and for AIP download there is no event as that doesn't require updates in the dashboard.

I'd consider setting up a different event.Service for the new auditlog Event type and publishing different events based on the audit log needs.

jraddaoui replied:
Given this comment and the current implementation logging to a file per service, I'd also consider just passing the logger directly to the API service.

jraddaoui commented on enduro.toml:
```diff
@@ -80,6 +80,23 @@ migrate = true
+# compress determines if the rotated log files are compressed using gzip
+# (default: false).
+compress = false
+# verbosity determines the minimum level threshold for a log message to be
+# recorded. Currently supported levels are: -4 = DEBUG, 0 = INFO (default),
+# 4 = WARN, 8 = ERROR.
+verbosity = 0
```
Do we care about verbosity in an audit log? I don't think so, we want to log all audit events, I'd remove it from the config.

djjuhasz replied:
Why would you assume that everyone using Enduro would want to log all audit events?

jraddaoui replied:
Yeah, you are right, I'm not sure if verbosity is the way to filter them, but I can't think of a better option. For now I think it is unnecessary though.

jraddaoui commented on internal/auditlog/config.go:
```diff
@@ -0,0 +1,57 @@
+	}
+
+	logger := slog.New(slog.NewJSONHandler(w, &slog.HandlerOptions{
+		Level: slog.Level(cfg.Verbosity),
```
Whatever is the default level, use it in all events. And no need to have it in the config or the event type.

djjuhasz replied:
In my research on audit logs it seemed like common practice to include error levels and to log events like failed logins, which would be a WARN or ERROR level event. I know we are only using INFO right now, but I think we should build on the assumption that we will have different event levels in the future.

jraddaoui replied:
I'd think of that as the outcome of the action, which we are not recording yet. I think it's quite confusing this way, we'd need to represent somehow the levels of each event. What I read is that we should make a distinction between blocking events (those that MUST be logged) and non-blocking events (those not that relevant that may be skipped based on the system load). I think we shouldn't worry about that with the current implementation, and I think the same about the verbosity.

For now I think it adds some confusion and it's something that we don't need to make configurable, I may be proven wrong, but we could easily add it later if that's the case.

jraddaoui commented on internal/auditlog/event.go:
```diff
@@ -0,0 +1,19 @@
+import "log/slog"
+
+type Event struct {
+	Level    slog.Level
```
Again, I don't think we care about this here.

jraddaoui commented on internal/auditlog/event.go:
```diff
@@ -0,0 +1,19 @@
+	Msg      string
+	Type     string
+	ObjectID string
+	UserID   string
```
I think the user id is not enough in this logs, at least I'd like to see the user email in the audit events. Otherwise it doesn't provide a lot of info as a log, still requires to go to the DB to get the actual user details.

djjuhasz replied:
Yeah, I agree we should log user emails, especially since we don't have a user management UI.

jraddaoui commented on internal/ingest/convert.go:
```diff
@@ -177,3 +178,22 @@ func sipSourceObjectsToGoa(objects []*sipsource.Object) goaingest.SIPSourceObjec
 	return r
 }
+
+func HandleAuditEvent(ev *goaingest.IngestEvent) *auditlog.Event {
+	switch e := ev.IngestValue.(type) {
+	case *goaingest.SIPCreatedEvent:
+		var userID string
+		if e.Item.UploaderUUID != nil {
+			userID = e.Item.UploaderUUID.String()
+		}
+		return &auditlog.Event{
+			Msg:      "SIP created",
+			Type:     "sip.created",
+			ObjectID: e.UUID.String(),
+			UserID:   userID,
+		}
+	default:
+		// Ignore unsupported event types.
+		return nil
+	}
+}
```
Why do we have both `Type` and `Msg` fields? I think the more human readable version would be enough. To be clear, I'd use the `Msg` value as `Type` and remove the `Msg` field.

djjuhasz replied:
@fiver-watson suggested a dot separated type vocabulary (e.g. `SIP.uploaded`) for the audit log events in the original description of #1282. I think a more discursive event message is better for human readers. I'm fine with eliminating the `Type`, but I'd be more reluctant to removing the human readable `Msg`.

jraddaoui replied:
I see, and based on our discussion we could have different messages based on the outcome of the action, while the type would be the same, so I'm cool with having both. One minor detail, should it be "sip.creation" as the type then?

djjuhasz replied:
Yes, changing the type to "SIP.creation" is a good idea. :+1:

fiver-watson replied:
Also curious about "SIP creation" as an audit log event in general.... perhaps from the perspective of the Enduro DB we are "creating" the SIP when first registering it, but otherwise, I'm not sure how we end up with a SIP creation event in Enduro when the application doesn't support SIP creation, and producers generally do this before content ever reaches Enduro...?

If we are thinking of the staging locations (where ingest doesn't proceed immediately for example), I'd say "SIP deposited" would be a better event, and then perhaps an "Ingest started" or "SIP ingest started" etc event when things are launched? Created doesn't feel quite right... but I also might not be understanding the context here, so apologies in advance!

jraddaoui commented on internal/auditlog/event.go:
```diff
@@ -0,0 +1,19 @@
+	Level    slog.Level
+	Msg      string
+	Type     string
+	ObjectID string
```
That looks so AtoM and OOP :smile:, maybe "ResourceID"? Not a big deal of course.

jraddaoui reviewed (commented):
Thanks @djjuhasz! See some comments below.

About the `defer` issue we were talking in the meeting and using the group actors. Wouldn't it be enough to use the main/run pattern we use in other projects? Calling `run` from `main` and if there is an error calling `os.Exit(1)`, `run` would execute the deferred functions before returning the error.

Also, what's the plan for storage? A different or the same file? We don't have internal user IDs there, and even if we add them, they will be different from ingest.

djjuhasz commented:
We implemented a simpler audit logger implementation in commit 6dba5179ed6f2b0c571b842d9b17092c43165b35 that doesn't require subscribing to the event service and emitting events to the Monitor API that are not required for the Dashboard.
